pyrit/prompt_target/azure_ml_chat_target.py

```diff
@@ -0,0 +1,217 @@
+        self,
+        *,
+        endpoint_uri: str = None,
+        api_key: str = None,
+        chat_message_normalizer: ChatMessageNormalizer = ChatMessageNop(),
+        memory: MemoryInterface = None,
+        max_tokens: int = 400,
+        temperature: float = 1.0,
+        top_p: int = 1,
+        repetition_penalty: float = 1.2,
+    ) -> None:
+        PromptChatTarget.__init__(self, memory=memory)
+
+        self.endpoint_uri: str = default_values.get_required_value(
+            env_var_name=self.ENDPOINT_URI_ENVIRONMENT_VARIABLE, passed_value=endpoint_uri
+        )
+        self.api_key: str = default_values.get_required_value(
+            env_var_name=self.API_KEY_ENVIRONMENT_VARIABLE, passed_value=api_key
+        )
+        self.chat_message_normalizer = chat_message_normalizer
+
+        self._max_tokens = max_tokens
+        self._temperature = temperature
+        self._top_p = top_p
+        self._repetition_penalty = repetition_penalty
+
+    def set_system_prompt(self, *, prompt: str, conversation_id: str, normalizer_id: str) -> None:
+        messages = self.memory.get_memories_with_conversation_id(conversation_id=conversation_id)
+
+        if messages:
+            raise RuntimeError("Conversation already exists, system prompt needs to be set at the beginning")
+
+        self.memory.add_chat_message_to_memory(
+            conversation=ChatMessage(role="system", content=prompt),
+            conversation_id=conversation_id,
+            normalizer_id=normalizer_id,
+        )
+
+    def send_prompt(self, *, normalized_prompt: str, conversation_id: str, normalizer_id: str) -> str:
+        messages = self._prepare_message(normalized_prompt, conversation_id, normalizer_id)
+
+        resp = self.complete_chat(
+            messages=messages,
+            temperature=self._temperature,
+            top_p=self._top_p,
+            repetition_penalty=self._repetition_penalty
+        )
+
+        self.memory.add_chat_message_to_memory(
+            ChatMessage(role="assistant", content=resp),
+            conversation_id=conversation_id,
+            normalizer_id=normalizer_id,
+        )
+
+        return resp
+
+    async def send_prompt_async(self, *, normalized_prompt: str, conversation_id: str, normalizer_id: str) -> str:
+        messages = self._prepare_message(normalized_prompt, conversation_id, normalizer_id)
+
+        resp = await self.complete_chat_async(
+                    messages=messages,
+                    temperature=self._temperature,
+                    top_p=self._top_p,
+                    repetition_penalty=self._repetition_penalty
+        )
+
+        self.memory.add_chat_message_to_memory(
+            ChatMessage(role="assistant", content=resp),
+            conversation_id=conversation_id,
+            normalizer_id=normalizer_id,
+        )
+
+        return resp
+
+    def complete_chat(
+        self,
+        messages: list[ChatMessage],
+        max_tokens: int = 400,
```

Comment:
```suggestion
        max_tokens: Optional[int] = 400,
```
Might have to import from typing

Reply:
Same in other cases

Reply:
hmmmmm, seems redundant to use optional when there is a default value. ChatGPT gave a response I like

In summary:

Use Optional if None is a meaningful, distinct case that your function is designed to handle.
Do not use Optional if None is not an expected or meaningful value for the function, especially if the parameters have sensible defaults and are intended to always have a value that matches their type hint.

Reply:
Huh, perhaps I misunderstood it's meaning so far! I suppose you're right. We don't actually want `None` here. So if it was
`max_tokens = None` as default then Optional[int] would have been right. Hey, at least I learned something here 😆